README.md
# BindingPrediction
Predict binding sites of proteins on DNA using CNNs.


ChangeLog
2016-03-21  Deniz Yuret  <dyuret@ku.edu.tr>

	* model05.jl: trying rnn model.  saw 62.6@4K with one lstm 32
	hidden.  try without adam?
	with 256 hidden: 64.2@8K, 65.8@16K, 66.95@32K, 67.95@64K
	(1,0.7670895f0,0.5)
	(2,0.7870218932628632,0.5)
	(4,0.790457211508155,0.5)
	(8,0.7887824499360889,0.4999249843750001)
	(16,0.7820475894107846,0.5017376894477075)
	(32,0.7674567686487525,0.5133441129336902)
	(64,0.7398459320816413,0.5397198113865014)
	(128,0.6877308620807926,0.5708688192357589)
	(256,0.6054717233889524,0.5938594454467707)
	(512,0.5693346123727129,0.608386887163344)
	(1024,0.5630231562766616,0.6128376101608914)
	(2048,0.5585068162888615,0.6111671552820388)
	(4096,0.5593190887857925,0.6195320647954465)
	(8192,0.5487005717722863,0.6418676054273353)
	(16384,0.5409262118008304,0.6578520411961633)
	(32768,0.5287464149346305,0.669545893339313)
	(65536,0.5204879012412502,0.6795305601652457)
	(131072,0.5232149835850248,0.6751933279964852)
	(262144,0.5247984440994684,0.6786203322967537)

2016-03-20  Deniz Yuret  <dyuret@ku.edu.tr>

	* model04.jl: vggnet model A from http://arxiv.org/pdf/1409.1556v6.pdf
	62@1K

	* model03.jl: more fc layers: add +.5% => 66.2@8K
	C=64,128,256
	F=256 %64@512 %65.4@1K 66.2@8K
	F=128 and 512 are similar
	F=128,128 %64.3@512 %65.6@1K 66.2@8K

	Longer run: (iter,softloss,1-error) with nbatch=128
	(16,0.7717592019118032,0.5067542147176836)
	(32,0.7471922323210548,0.521499697094878)
	(64,0.6986527361455038,0.5516034758784856)
	(128,0.6340780189803857,0.5882834592249624)
	(256,0.5712480153757631,0.6298053340609724)
	(512,0.5517600432731254,0.6417776660928463)
	(1024,0.5452585118039903,0.6530180761610829)
	(2048,0.5409149309070316,0.6581438928103153)
	(4096,0.5408834846403274,0.6582771821489561)
	(8192,0.5369208570946014,0.6645235420027924)
	(16384,0.5334389944026348,0.670336899997459)
	(32768,0.5288384511978679,0.6688833300227164)
	(65536,0.5205062706206457,0.6781443354108854)
	(131072,0.5238383187543814,0.6748884933349375)
	(262144,0.5267178587417906,0.6747090959623597)

	* model02.jl: more conv layers: add +1% => 65.8@8K
	C=64,128 %64@1K %65@8K
	C=128,256 similar
	C=64,128,256 %65@1K %65.8@8K
	C=64,128,256,512 %62.4@256 %63.8@512 %65.0@1K

	* model01.jl: initial experiments with a cbfp+wbf: 64.7@32K

	%63@1024 C=512 W=3 B=128
	crosses %64.7 @32K
	C=256 or C=512 not better, W=5 similar, tanh similar

	* TODO:
	# gclip
	# more conv layers
	# more fc layers
	# embedding layer
	# tanh vs relu

	* chars: there is N in addition to ACTG:

	* wc: each sequence is 51 long.  approx 25M positive 25M negative
	examples.  17M unique in each case.

	dyuret@biyofiz-4-0:~/alkan[0]$ zcat NeuroD2\ ChIP-SEQ-antibody1.fastq.gz | awk 'NR%4==2' | wc -l
	26767599
	dyuret@biyofiz-4-0:~/alkan[35]$ zcat GFP\ ChIP-SEQ-antibody1.fastq.gz | awk 'NR%4==2' | wc -l
	24495706
	dyuret@biyofiz-4-0:~/alkan[96]$ zcat NeuroD2\ ChIP-SEQ-antibody1.fastq.gz | awk 'NR%4==2' | rcount | wc -l
	17483930
	dyuret@biyofiz-4-0:~/alkan[0]$ zcat GFP\ ChIP-SEQ-antibody1.fastq.gz | awk 'NR%4==2' | rcount | wc -l
	17615277

	dyuret@biyofiz-4-0:~/alkan[0]$ zcat NeuroD2\ ChIP-SEQ-antibody1.fastq.gz | awk 'NR%4==2' | sort -u > foo.neuro.sortu
	dyuret@biyofiz-4-0:~/alkan[213]$ zcat GFP\ ChIP-SEQ-antibody1.fastq.gz  | awk 'NR%4==2' | sort -u > foo.gfp.sortu
	dyuret@biyofiz-4-0:~/alkan[182]$ wc -l *.sortu
	17615277 foo.gfp.sortu
	17483930 foo.neuro.sortu
	35099207 total
	dyuret@biyofiz-4-0:~/alkan[1]$ comm -12 *.sortu | wc -l
	553500


model01.jl
using Knet
include("readdata.jl")

# Try a convolutional model, similar to CharCNN in http://arxiv.org/pdf/1508.06615.pdf
# The input is a sequence of characters of length T.
# We map this to an embedding matrix DxT (alternative is to use one-hot with D=4).
# We apply a 1-D convolution bank (width and number of filters to be decided).
# In the simplest case with one conv layer, we apply max pooling of window T.
# We apply an affine transform and a softmax to the output.
# B shows the minibatch size, and everything is in 4D for CUDNN ops.

# x is a (1,T,D,N) matrix of inputs where T is the sequence length, D is the dimension of each char embedding, N is the minibatch size
# we convolve with a (1,W,D,C) filter bank with W/2 padding to avoid dimension change
# this gives a (1,T,C,N) output which gets a bias added and f applied, neither of which effects the dimension
# we apply max pooling with pwindow=T which gives an output of (1,1,C,N)
# we have a fully connected layer ending with a softmax at the end
# cbfp73 has the following default options each of which can be modified in model01:
# cwindow=3, cpadding=div(cwindow,2), cstride=1, cmode=Knet.CUDNN_CONVOLUTION,
# pwindow=2, ppadding=0, pstride=pwindow, pmode=Knet.CUDNN_POOLING_MAX,
# cinit=Xavier(), binit=Constant(0), f=:relu

@knet function model01(x; T=51, D=4, C=512, W=3, o...)
    y = cbfp73(x; o..., pwindow=T, out=C)
    return wbf73(y; o..., out=2, f=:soft)
end

function train01(; N=2^13, lr=0.001, adam=true, nbatch=128, o...)
    global f = compile(:model01; o...)
    setp(f, lr=lr, adam=adam)
    global d = GulayseData(nbatch)
    sloss = zloss = -1; n = nextn = 1
    for (x,ygold) in d
        ypred = forw(f, x)
        sl = softloss(ypred,ygold); sloss = (n==1 ? sl : 0.99 * sloss + 0.01 * sl)
        zl = zeroone(ypred,ygold);  zloss = (n==1 ? zl : 0.99 * zloss + 0.01 * zl)
        n==nextn && (println((n,sloss,1-zloss)); nextn*=2)
        back(f, ygold, softloss)
        update!(f)
        (n += 1) > N && break
    end
end

train01()


model02.jl
using Knet
include("readdata.jl")

# Similar to model01, trying more convolutional layers

@knet function model02(x; T=51, D=4, W=3, o...)
    a = cbf73(x; o..., out=64)
    b = cbf73(a; o..., out=128)
    c = cbf73(b; o..., out=256)
    y = pool(c; o..., window=T)
    return wbf73(y; o..., out=2, f=:soft)
end

function train02(; N=2^13, lr=0.001, adam=true, nbatch=128, o...)
    global f = compile(:model02; o...)
    setp(f, lr=lr, adam=adam)
    global d = GulayseData(nbatch)
    sloss = zloss = -1; n = nextn = 1
    for (x,ygold) in d
        ypred = forw(f, x)
        sl = softloss(ypred,ygold); sloss = (n==1 ? sl : 0.99 * sloss + 0.01 * sl)
        zl = zeroone(ypred,ygold);  zloss = (n==1 ? zl : 0.99 * zloss + 0.01 * zl)
        n==nextn && (println((n,sloss,1-zloss)); nextn*=2)
        back(f, ygold, softloss)
        update!(f)
        (n += 1) > N && break
    end
end

train02()


model03.jl
using Knet
include("readdata.jl")

# Similar to model02, trying more fully connected layers

@knet function model03(x; T=51, D=4, W=3, o...)
    c1 = cbf73(x; o..., out=64)
    c2 = cbf73(c1; o..., out=128)
    c3 = cbf73(c2; o..., out=256)
    c = pool(c3; o..., window=T)
    f1 = wbf73(c; o..., out=128)
    f2 = wbf73(f1; o..., out=128)
    return wbf73(f2; o..., out=2, f=:soft)
end

function train03(; N=2^13, lr=0.001, adam=true, nbatch=128, o...)
    global f = compile(:model03; o...)
    setp(f, lr=lr, adam=adam)
    global d = GulayseData(nbatch)
    sloss = zloss = -1; n = nextn = 1
    for (x,ygold) in d
        ypred = forw(f, x)
        sl = softloss(ypred,ygold); sloss = (n==1 ? sl : 0.99 * sloss + 0.01 * sl)
        zl = zeroone(ypred,ygold);  zloss = (n==1 ? zl : 0.99 * zloss + 0.01 * zl)
        n==nextn && (println((n,sloss,1-zloss)); nextn*=2)
        back(f, ygold, softloss)
        update!(f)
        (n += 1) > N && break
    end
end

train03()


model05.jl
using Knet
using Knet: stack_isempty
include("readdata.jl")

# Trying an rnn model.

@knet function model05(x; hidden=0, o...)
    h = lstm(x; o..., out=hidden)
    if predict
        return wbf(h; o..., out=2, f=:soft)
    end
end

function rnnforw(f, x)
    # x is a (1,T,D,N) one-hot tensor
    # T:sequence length, D:input dimension, N:minibatch size
    global xdbg = x
    (X,T,D,N) = size(x)
    reset!(f)
    ypred = nothing
    for t=1:T
        xt = reshape(x[1,t,:,:], (D, N))
        ypred = sforw(f, xt; predict=(t==T))
    end
    return ypred
end

function train05(; N=2^10, lr=0.001, adam=true, nbatch=128, gclip=0, hidden=256, seed=0, o...)
    seed > 0 && setseed(seed)
    global f = compile(:model05; hidden=hidden, o...)
    setp(f, lr=lr, adam=adam)
    global d = GulayseData(nbatch)
    sloss = zloss = -1; n = nextn = 1
    for (x,ygold) in d
        ypred = rnnforw(f, x)
        sback(f, ygold, softloss)
        while !stack_isempty(f); sback(f); end
        gscale = (gclip > 0 ? min(1.0, gclip/gnorm(f)) : 1.0)
        update!(f; gscale=gscale)
        sl = softloss(ypred,ygold); sloss = (n==1 ? sl : 0.99 * sloss + 0.01 * sl)
        zl = zeroone(ypred,ygold);  zloss = (n==1 ? zl : 0.99 * zloss + 0.01 * zl)
        n==nextn && (println((n,sloss,1-zloss)); nextn*=2)
        (n += 1) > N && break
    end
end

train05()


readdata.jl
using GZip
const file0 = "GFP ChIP-SEQ-antibody1.fastq.gz"
const file1 = "NeuroD2 ChIP-SEQ-antibody1.fastq.gz"

# The files are too big, we'll just write an iterator that reads from
# the file as needed.

import Base: start, done, next

type GulayseData
    nbatch::Int
    flip::Bool
    f0::GZipStream
    f1::GZipStream
    x::Array{Float32,4}
    y::Array{Float32,2}
    GulayseData(n=128,f=true)=new(n,f)
end

function start(d::GulayseData)
    isdefined(d,:f0) && close(f0)
    isdefined(d,:f1) && close(f1)
    d.f0 = GZip.open(file0)
    d.f1 = GZip.open(file1)
    d.x = zeros(Float32, 1, 51, 4, d.nbatch)
    d.y = zeros(Float32, 2, d.nbatch)
    return nothing
end

function done(d::GulayseData, s)
    eof(d.f0) || eof(d.f1)
end

function next(d::GulayseData, s)
    d.x[:] = d.y[:] = 0
    for n=1:d.nbatch      # how do we handle train vs test?
        d.flip ? d.y[1,n]=1 : d.y[2,n]=1 # do not have a loss function for 1-D output
        xread(d.flip ? d.f1 : d.f0, d.x, n) # alternate reading from positive and negative examples
        d.flip = !d.flip
    end
    return ((d.x, d.y), nothing)
end

function xread(fp::GZipStream, x::Array{Float32,4}, n::Int, s::ASCIIString="")
    for i=1:4
        eof(fp) && return
        tmp = readline(fp)
        i==2 && (s = tmp)
    end
    length(s)==52 || error("length(s)==$(length(s))")
    for i=1:51
        s[i] == 'A' ? x[1,i,1,n] = 1 :
        s[i] == 'C' ? x[1,i,2,n] = 1 :
        s[i] == 'G' ? x[1,i,3,n] = 1 :
        s[i] == 'T' ? x[1,i,4,n] = 1 :
        s[i] == 'N' ? nothing :
        error("s[$i]==$(s[i])")
    end
end


model04.jl
using Knet
include("readdata.jl")

# Copy of VGG model A: http://arxiv.org/pdf/1409.1556v6.pdf

@knet function model04(x; T=51, D=4, W=3, o...)
    c1 = cbfp73(x;  o..., out=64, pwindow=T)
    c2 = cbfp73(c1; o..., out=128, pwindow=T)
    c3 = cbf73(c2; o..., out=256)
    c4 = cbfp73(c3; o..., out=256, pwindow=T)
    c5 = cbf73(c4; o..., out=512)
    c6 = cbfp73(c5; o..., out=512, pwindow=T)
    c7 = cbf73(c6; o..., out=512)
    c8 = cbfp73(c7; o..., out=512, pwindow=T)
    f1 = wbf73(c8; o..., out=4096)
    f2 = wbf73(f1; o..., out=4096)
    return wbf73(f2; o..., out=2, f=:soft)
end

function train04(; N=2^13, lr=0.001, adam=true, nbatch=128, o...)
    global f = compile(:model04; o...)
    setp(f, lr=lr, adam=adam)
    global d = GulayseData(nbatch)
    sloss = zloss = -1; n = nextn = 1
    for (x,ygold) in d
        ypred = forw(f, x)
        sl = softloss(ypred,ygold); sloss = (n==1 ? sl : 0.99 * sloss + 0.01 * sl)
        zl = zeroone(ypred,ygold);  zloss = (n==1 ? zl : 0.99 * zloss + 0.01 * zl)
        n==nextn && (println((n,sloss,1-zloss)); nextn*=2)
        back(f, ygold, softloss)
        update!(f)
        (n += 1) > N && break
    end
end

train04()
